marconi-sidechain/test/Spec/Marconi/Sidechain/Utils.hs

```diff
@@ -0,0 +1,155 @@
+    IO.putMVar mvar contents
+  IO.takeMVar mvar
+
+-- -- | Compute the path to the binary given a package name or an environment variable override.
+binFlex
+  :: String
+  -- ^ Package name
+  -> String
+  -- ^ Environment variable pointing to the binary to run
+  -> IO FilePath
+  -- ^ Path to executable
+binFlex pkg binaryEnv = do
+  maybeEnvBin <- liftIO $ IO.lookupEnv binaryEnv
+  case maybeEnvBin of
+    Just envBin -> return envBin
+    Nothing -> binDist pkg
+
+addExeSuffix :: String -> String
+addExeSuffix s =
+  if ".exe" `L.isSuffixOf` s
+    then s
+    else s <> if OS.isWin32 then ".exe" else ""
+
+-- -- | Find the nearest plan.json going upwards from the current directory.
+findDefaultPlanJsonFile :: IO FilePath
+findDefaultPlanJsonFile = IO.getCurrentDirectory >>= go
+  where
+    go :: FilePath -> IO FilePath
+    go d = do
+      let file = d </> "dist-newstyle/cache/plan.json"
+      exists <- IO.doesFileExist file
+      if exists
+        then return file
+        else do
+          let parent = IO.takeDirectory d
+          if parent == d
+            then return "dist-newstyle/cache/plan.json"
+            else go parent
+
+-- -- | Discover the location of the plan.json file.
+planJsonFile :: String
+planJsonFile = IO.unsafePerformIO $ do
+  maybeBuildDir <- liftIO $ IO.lookupEnv "CABAL_BUILDDIR"
+  case maybeBuildDir of
+    Just buildDir -> return $ ".." </> buildDir </> "cache/plan.json"
+    Nothing -> findDefaultPlanJsonFile
+
+-- -- | Consult the "plan.json" generated by cabal to get the path to the executable corresponding.
+-- -- to a haskell package.  It is assumed that the project has already been configured and the
+-- -- executable has been built.
+binDist
+  :: String
+  -- ^ Package name
+  -> IO FilePath
+  -- ^ Path to executable
+binDist pkg = do
+  contents <- BSL.readFile planJsonFile
+
+  case A.eitherDecode contents of
+    Right plan -> case L.filter matching (plan & H.installPlan) of
+      (component : _) -> case component & H.binFile of
+        Just bin -> return $ addExeSuffix (T.unpack bin)
+        Nothing -> error $ "missing bin-file in: " <> show component
+      [] -> error $ "Cannot find exe:" <> pkg <> " in plan"
+    Left message -> error $ "Cannot decode plan: " <> message
+  where
+    matching :: H.Component -> Bool
+    matching component = case H.componentName component of
+      Just name -> name == "exe:" <> T.pack pkg
+      Nothing -> False
+
+{- | Create a 'CreateProcess' describing how to start a process given the Cabal package name
+ corresponding to the executable, an environment variable pointing to the executable,
+ and an argument list.
+
+ The actual executable used will the one specified by the environment variable, but if
+ the environment variable is not defined, it will be found instead by consulting the
+ "plan.json" generated by cabal.  It is assumed that the project has already been
+ configured and the executable has been built.
+-}
+procFlex
```

Comment:
Maybe a better name? I don't know what "flex" refers to.

Reply:
I can improve the description. 'flex' means that the environment determines how the process is launched:

When the environment variable is set, for us this means in the nix ci environment) the 'envBin' argument describes the environment variable that defines the binary to use to launch the process.

When running outside a nix environment, the `pkgBin` describes the name of the binary to launch via cabal exec.

Reply:
Yep, an improved description works for me

Reply:
Done